diff -ruP latest/drivers/usb/host/xhci.c working/drivers/usb/host/xhci.c
--- latest/drivers/usb/host/xhci.c	2022-03-08 06:55:56.000000000 -0600
+++ working/drivers/usb/host/xhci.c	2022-02-28 10:31:35.000000000 -0600
@@ -1091,7 +1091,6 @@
 	int			retval = 0;
 	bool			comp_timer_running = false;
 	bool			pending_portevent = false;
-	bool			reinit_xhc = false;
 
 	if (!hcd->state)
 		return 0;
@@ -1108,11 +1107,10 @@
 	set_bit(HCD_FLAG_HW_ACCESSIBLE, &xhci->shared_hcd->flags);
 
 	spin_lock_irq(&xhci->lock);
+	if ((xhci->quirks & XHCI_RESET_ON_RESUME) || xhci->broken_suspend)
+		hibernated = true;
 
-	if (hibernated || xhci->quirks & XHCI_RESET_ON_RESUME || xhci->broken_suspend)
-		reinit_xhc = true;
-
-	if (!reinit_xhc) {
+	if (!hibernated) {
 		/*
 		 * Some controllers might lose power during suspend, so wait
 		 * for controller not ready bit to clear, just as in xHC init.
@@ -1145,17 +1143,12 @@
 			spin_unlock_irq(&xhci->lock);
 			return -ETIMEDOUT;
 		}
+		temp = readl(&xhci->op_regs->status);
 	}
 
-	temp = readl(&xhci->op_regs->status);
-
-	/* re-initialize the HC on Restore Error, or Host Controller Error */
-	if (temp & (STS_SRE | STS_HCE)) {
-		reinit_xhc = true;
-		xhci_warn(xhci, "xHC error in resume, USBSTS 0x%x, Reinit\n", temp);
-	}
+	/* If restore operation fails, re-initialize the HC during resume */
+	if ((temp & STS_SRE) || hibernated) {
 
-	if (reinit_xhc) {
 		if ((xhci->quirks & XHCI_COMP_MODE_QUIRK) &&
 				!(xhci_all_ports_seen_u0(xhci))) {
 			del_timer_sync(&xhci->comp_mode_recovery_timer);
@@ -1395,12 +1388,9 @@
 static int xhci_map_urb_for_dma(struct usb_hcd *hcd, struct urb *urb,
 				gfp_t mem_flags)
 {
-	unsigned int i, maxpacket;
-	struct scatterlist *sg;
 	struct xhci_hcd *xhci;
 
 	xhci = hcd_to_xhci(hcd);
-	maxpacket = usb_endpoint_maxp(&urb->ep->desc);
 
 	if (xhci_urb_suitable_for_idt(urb))
 		return 0;
@@ -1409,16 +1399,6 @@
 		if (xhci_urb_temp_buffer_required(hcd, urb))
 			return xhci_map_temp_buffer(hcd, urb);
 	}
-
-	if (xhci->quirks & XHCI_VLI_SS_BULK_OUT_BUG &&
-	    usb_endpoint_is_bulk_out(&urb->ep->desc) &&
-	    urb->dev->speed >= USB_SPEED_SUPER &&
-	    urb->transfer_buffer_length != 0) {
-		for_each_sg(urb->sg, sg, urb->num_sgs, i) {
-			if (sg->length % maxpacket)
-				return xhci_map_temp_buffer(hcd, urb);
-		}
-	}
 	return usb_hcd_map_urb_for_dma(hcd, urb, mem_flags);
 }
 
@@ -1432,8 +1412,7 @@
 	if (urb->num_sgs && (urb->transfer_flags & URB_DMA_MAP_SINGLE))
 		unmap_temp_buf = true;
 
-	if ((xhci->quirks & (XHCI_SG_TRB_CACHE_SIZE_QUIRK | XHCI_VLI_SS_BULK_OUT_BUG))
-	     && unmap_temp_buf)
+	if ((xhci->quirks & XHCI_SG_TRB_CACHE_SIZE_QUIRK) && unmap_temp_buf)
 		xhci_unmap_temp_buf(hcd, urb);
 	else
 		usb_hcd_unmap_urb_for_dma(hcd, urb);
@@ -1728,12 +1707,9 @@
 	struct urb_priv	*urb_priv;
 	int num_tds;
 
-	if (!urb)
+	if (!urb || xhci_check_args(hcd, urb->dev, urb->ep,
+					true, true, __func__) <= 0)
 		return -EINVAL;
-	ret = xhci_check_args(hcd, urb->dev, urb->ep,
-					true, true, __func__);
-	if (ret <= 0)
-		return ret ? ret : -EINVAL;
 
 	slot_id = urb->dev->slot_id;
 	ep_index = xhci_get_endpoint_index(&urb->ep->desc);
@@ -3450,7 +3426,7 @@
 		return -EINVAL;
 	ret = xhci_check_args(xhci_to_hcd(xhci), udev, ep, 1, true, __func__);
 	if (ret <= 0)
-		return ret ? ret : -EINVAL;
+		return -EINVAL;
 	if (usb_ss_max_streams(&ep->ss_ep_comp) == 0) {
 		xhci_warn(xhci, "WARN: SuperSpeed Endpoint Companion"
 				" descriptor for ep 0x%x does not support streams\n",
diff -ruP latest/drivers/usb/host/xhci-ring.c working/drivers/usb/host/xhci-ring.c
--- latest/drivers/usb/host/xhci-ring.c	2022-03-08 06:55:56.000000000 -0600
+++ working/drivers/usb/host/xhci-ring.c	2022-02-28 10:31:35.000000000 -0600
@@ -3607,7 +3607,7 @@
 	unsigned int num_trbs;
 	unsigned int start_cycle, num_sgs = 0;
 	unsigned int enqd_len, block_len, trb_buff_len, full_len;
-	int sent_len, ret, vli_bulk_quirk = 0;
+	int sent_len, ret, vli_quirk = 0;
 	u32 field, length_field, remainder, maxpacket;
 	u64 addr, send_addr;
 
@@ -3653,9 +3653,14 @@
 	send_addr = addr;
 
 	if (xhci->quirks & XHCI_VLI_SS_BULK_OUT_BUG &&
-	    usb_endpoint_is_bulk_out(&urb->ep->desc)
-	    && urb->dev->speed >= USB_SPEED_SUPER) {
-		vli_bulk_quirk = 1;
+	    !usb_urb_dir_in(urb) && urb->dev->speed >= USB_SPEED_SUPER) {
+		/*
+		 * VL805 - superspeed bulk OUT traffic can cause
+		 * an internal fifo overflow if the TRB buffer is larger
+		 * than wMaxPacket and the length is not an integer
+		 * multiple of wMaxPacket.
+		 */
+		vli_quirk = 1;
 	}
 
 	/* Queue the TRBs, even if they are zero-length */
@@ -3670,7 +3675,7 @@
 		if (enqd_len + trb_buff_len > full_len)
 			trb_buff_len = full_len - enqd_len;
 
-		if (vli_bulk_quirk && trb_buff_len > maxpacket) {
+		if (vli_quirk && trb_buff_len > maxpacket) {
 			/* SS bulk wMaxPacket is 1024B */
 			remainder = trb_buff_len & (maxpacket - 1);
 			trb_buff_len -= remainder;
